File: csvtransform.py
import pandas as pd

sample1 = pd.read_csv("sample1.csv")
sample2 = pd.read_csv("sample2.csv")

def fastmerge(d1, d2):
    d1_names = d1.columns
    d2_names = d2.columns
    
    # columns in d1 but not in d2
    d2_add = list(set(d1_names) - set(d2_names))
    
    # columns in d2 but not in d1
    d1_add = list(set(d2_names) - set(d1_names))
    
    # add blank columns to d2
    if d2_add:
        for col in d2_add:
            d2[col] = pd.NA
    
    # add blank columns to d1
    if d1_add:
        for col in d1_add:
            d1[col] = pd.NA
    
    return pd.concat([d1, d2], ignore_index=True)

table = fastmerge(sample1, sample2)

# Replace null values with 0's
table.fillna(0, inplace=True)

# Extract year and month from the date column
table['year'] = table['date'].str[-4:]
table['month'] = table['date'].str[5:7]

# Create a new key column by concatenating year and month
table['key'] = table['year'] + "-" + table['month']

# Remove spaces from the key column
table['key'] = table['key'].str.replace(" ", "")

# Drop the month and year columns
table.drop(['month', 'year'], axis=1, inplace=True)

# Write the table to a CSV file
table.to_csv("splunkcloudingest.csv", index=True, quoting=pd.QUOTE_NONE)

File: csvtransform.R

files <- list.files(path = "data/")
f <- list()

sample1 <- read.csv(paste0("data/",files[1]), header=TRUE, sep=",")
sample2 <- read.csv(paste0("data/",files[2]), header=TRUE, sep=",")

fastmerge <- function(d1, d2) {
  d1.names <- names(d1)
  d2.names <- names(d2)
  
  # columns in d1 but not in d2
  d2.add <- setdiff(d1.names, d2.names)
  
  # columns in d2 but not in d1
  d1.add <- setdiff(d2.names, d1.names)
  
  # add blank columns to d2
  if(length(d2.add) > 0) {
    for(i in 1:length(d2.add)) {
      d2[d2.add[i]] <- NA
    }
  }
  
  # add blank columns to d1
  if(length(d1.add) > 0) {
    for(i in 1:length(d1.add)) {
      d1[d1.add[i]] <- NA
    }
  }
  
  return(rbind(d1, d2))
}

table<- fastmerge(sample1, sample2)
for (i in 3:length(files)) {
  print(paste0("loading..",files[i]))
  sample2 <- read.csv(paste0("data/",files[i]), header=TRUE, sep=",")
  table<- fastmerge(table, sample2)
}

table[is.na(table)] <- 0     # replace nulls with 0's'

table$year<-substring(table$X_time,nchar(table$X_time)-3)
table$month<-substring(table$X_time,first=4,last=7)
table$key<-paste(table$year,"-", table$month)
table$key<-gsub("00:00:00", "", table$key)
table$month<-NULL
table$year<-NULL
table$X_time<-table$key
table$key<-NULL
table$limit<-NULL

ncols<-length(table)
table<-aggregate(table[2:ncols], by=list(date=table$X_time), FUN=mean)

#rownames(table) <- table[,1] # set rownames to col1 (dates)
#table[,1]<-NULL              # remove col1 (dates)

write.csv(table,file="transformed.csv", row.names = FALSE, quote=FALSE)

# table<-t(table)             # transpose table
# 
# index<-names(table)         # set vector of index names 
# values<-table               # set data values 
# dates<-rownames(table)      # set vector of dates
# 
# barplot(as.matrix(values), names.arg = index, xlab= "Date", ylab= "GB Ingest", 
#         legend.text = dates,
#         beside = FALSE) 


